Automated test case for "istioctl analyze" (#7781)
istio/istio.io: content/en/docs/ops/diagnostic-tools/istioctl-analyze/index.md

You can analyze your current live Kubernetes cluster by running:

{{< text syntax=bash snip_id=analyze_all_namespaces >}}
$ istioctl analyze --all-namespaces
{{< /text >}}

And that’s it! It’ll give you any recommendations that apply.

For example, if you forgot to enable Istio injection (a very common issue), you would get the following warning:

{{< text syntax=plain snip_id=analyze_all_namespace_sample_response >}}
...
Warn [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection
{{< /text >}}

> Fix the issue and try again:
> 
> {{< text syntax=bash snip_id=fix_default_namespace >}}
> $ kubectl label namespace default istio-injection=enabled --overwrite
> $ istioctl analyze --namespace default
> ✔ No validation issues found when analyzing namespace: default.
> {{< /text >}}

Reviewer: The --overwrite is an example of where somebody reading the doc would say "why do I need to do that?". It doesn't matter if it fails or not if running it on the command line, it's only a problem for the test. So, rather than complicating the doc to accommodate the test, it's better to split this into two text blocks.
```suggestion
Fix the issue:

{{< text syntax=bash snip_id=fix_default_namespace >}}
$ kubectl label namespace default istio-injection=enabled
{{< /text >}}

Then try again:

{{< text syntax=bash snip_id=try_with_fixed_namespace >}}
$ istioctl analyze --namespace default
✔ No validation issues found when analyzing namespace: default.
{{< /text >}}
```

You can analyze your current live Kubernetes cluster by running:

{{< text syntax=bash snip_id=analyze_all_namespaces >}}
$ istioctl analyze --all-namespaces
{{< /text >}}

And that’s it! It’ll give you any recommendations that apply.

For example, if you forgot to enable Istio injection (a very common issue), you would get the following warning:

{{< text syntax=plain snip_id=analyze_all_namespace_sample_response >}}

> ...
> Warn [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection

Reviewer: Using elision here has no advantage, because the text just says you'll see this message. Elision makes sense when you have a text block that shows the command and expected output, but you want to elide some of it.
```suggestion
Warn [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection
```

Author: But the rule output of `analyze all-namespaces` **MAY** include other messages, such as

```
Warn [IST0102] (Namespace ibm-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace ibm-system istio-injection=enabled' to enable it, or 'kubectl label namespace ibm-system istio-injection=disabled' to explicitly mark it as not needing injection
```

There is no way to predict what `--all-namespaces` will show for a given user.

Reviewer: That true, but doesn't matter. The text in the document isn't saying anything about whether or not there will be other output from the command, it's just saying that it there will be a warning like this if injection isn't enabled. For the test program, It's a perfect match for the `_verify_contains` function, which is just checking that the specified message is in the output. The elision is confusing, because the doc at this point doesn't (and shouldn't) mention anything about other possible errors that the command may find. To be perfectly safe for testing in the general case, you would actually need to add `...` both before and after the message of interest, but that's just unnecessary noise in the doc.

{{< /text >}}

Fix the issue and try again:

{{< text syntax=bash snip_id=fix_default_namespace >}}
$ kubectl label namespace default istio-injection=enabled --overwrite
$ istioctl analyze --namespace default
✔ No validation issues found when analyzing namespace: default.

> {{< /text >}}

Reviewer: The --overwrite is an example of where somebody reading the doc would say "why do I need to do that?". It doesn't matter if it fails or not if running it on the command line, it's only a problem for the test. So, rather than complicating the doc to accommodate the test, it's better to split this into two text blocks.
```suggestion
Fix the issue:

{{< text syntax=bash snip_id=fix_default_namespace >}}
$ kubectl label namespace default istio-injection=enabled --overwrite
{{< /text >}}

Then try again:

{{< text syntax=bash snip_id=try_with_fixed_namespace >}}
$ istioctl analyze --namespace default
✔ No validation issues found when analyzing namespace: default.
{{< /text >}}
```

{{< /text >}}

And that’s it! It’ll give you any recommendations that apply.

For example, if you forgot to enable Istio injection (a very common issue), you would get the following warning:

{{< text syntax=plain snip_id=analyze_all_namespace_sample_response >}}
Warn [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection
{{< /text >}}

Fix the issue:

{{< text syntax=bash snip_id=fix_default_namespace >}}
$ kubectl label namespace default istio-injection=enabled
{{< /text >}}

Then try again:

{{< text syntax=bash snip_id=try_with_fixed_namespace >}}
$ istioctl analyze --namespace default
✔ No validation issues found when analyzing namespace: default.
{{< /text >}}

## Analyzing live clusters, local files, or both

Analyze the current live cluster, simulating the effect of applying additional yaml files like `bookinfo-gateway.yaml` and `destination-rule-all.yaml` in the `samples/bookinfo/networking` directory:

{{< text syntax=bash snip_id=analyze_sample_destrule >}}

> $ istioctl analyze samples/bookinfo/networking/bookinfo-gateway.yaml samples/bookinfo/networking/destination-rule-all.yaml

Reviewer: YAML files were clickable links.

```suggestion
$ istioctl analyze @samples/bookinfo/networking/bookinfo-gateway.yaml@ @samples/bookinfo/networking/destination-rule-all.yaml@
```

Author: @frankbu They were!  I would like to bring that back.  I went through the _.md_ files and couldn't find any instances of `@samples` in `< text snip_id=... >` blocks.  Can you offer style guidance for this situation?

Reviewer: Most docs containing commands that use yaml files in the github samples dir wrap them in `@` chars so that the file is clickable in the UI. That includes this command (Look at line 36 in the diff, above, which has the `@` chars that you removed). https://preliminary.istio.io/latest/docs/ops/diagnostic-tools/istioctl-analyze/#analyzing-live-clusters-local-files-or-both. I'm just saying leave it as it was, don't change this in the doc.

For example. if you have a misconfigured gateway on your "ratings" virtual service, running `kubectl get virtualservice ratings` would give you something like:

{{< text syntax=yaml snip_id=vs_yaml_with_status >}}
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
...
  name: ratings
  namespace: default
...
spec:
  gateways:
  - bogus-gateway
  hosts:
  - ratings

> ...

Reviewer: Do you really need this many `...` in the example? It looks kind of ugly in the doc. Can't you cut it down to a couple and still get the point across?

Author: Maybe I don't understand `...` and elided.  For the user I thought it means "unimportant stuff" and for the recognizer I thought it means "*" (wildcard).  So for example I didn't want to show that there might be a second reported problem, so I used ... there.

Reviewer: You're right that `...` is a wildcard for testing and that in docs it's used to elide "unimportant stuff", but what you are doing is complicating the doc by trying to make the example a fully correct regular expression, which isn't necessary or helpful. In docs, you really just want to show a simple specific example, and only elide things that are noise (e.g., metadata or run specific values like date/time). The "unimportant stuff" you should elide is stuff in the original example that isn't needed to support what the document is trying to explain with the example. The original example didn't need to indicate the possible presence of other unrelated stuff that isn't relevant to the discussion at hand, so the elided one shouldn't either. Elision in docs is used to simplify, not to produce a generally correct regular expression.

Reviewer: I would leave the invalid specification in the doc, so it's clear what the error message is referring to.
```
kind: VirtualService
...
spec:
  gateways:
  - bogus-gateway
  hosts:
  - ratings
  http:
  - route:
    - destination:
        host: ratings
        subset: v1
status:
  validationMessages:
...
  - code: IST0101
    documentation_url: https://istio.io/docs/reference/config/analysis/IST0101?ref=status-controller
    level: Error
    message: 'Referenced gateway not found: "bogus-gateway"'
...
```

For example. if you have a misconfigured gateway on your "ratings" virtual service, running `kubectl get virtualservice ratings` would give you something like:

{{< text syntax=yaml snip_id=vs_yaml_with_status >}}
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
...
status:
  validationMessages:

> ...

Reviewer: Is this message coming out in the middle between other messages when you run the test? What other messages? If you can configure the test environment to only produce the one error, or to make it first or last, you could simplify the doc by removing one or both `...` around the messages.

You can also ignore specific analyzer messages using an annotation on the resource. For example, to ignore code IST0107 (`MisplacedAnnotation`) on resource `deployment/my-deployment`:

{{< text syntax=bash snip_id=annotate_for_deployment_suppression >}}
$ kubectl annotate deployment my-deployment galley.istio.io/analyze-suppress=IST0107
{{< /text >}}

To ignore multiple codes for a resource, separate each code with a comma:

{{< text syntax=bash snip_id=annotate_for_deployment_suppression_107 >}}

> $ kubectl annotate deployment my-deployment galley.istio.io/analyze-suppress=IST0107,IST0002 --overwrite

Reviewer: Instead of adding `--overwrite` in the doc, you should just remove the IST0107 suppress in the test.sh before calling this snip function. Since the doc describes this snip and the previous one as either or (as opposed to step by step run this, then run that), the reader might be confused why `--overwrite` is needed in the second snip but not the first. In general, you should try not to change the commands in the doc unless your fixing a bug or it makes something more clear for the reader.
```suggestion
$ kubectl annotate deployment my-deployment galley.istio.io/analyze-suppress=IST0107,IST0002
```
